src/resolver.rs

```diff
@@ -1353,7 +1353,38 @@ impl<'i> TypeAnnotator<'i> {
                                     .or_insert_with(std::vec::Vec::new)
                                     .push(candidate.to_string())
                             } else {
-                                params.push((parameter, type_name.to_string()))
+                                // intrinsic type promotion for variadics in order to be compatible with the C standard.
+                                // see ISO/IEC 9899:1999, 6.5.2.2 Function calls (https://www.open-std.org/jtc1/sc22/wg14/www/docs/n1256.pdf)
+                                // or https://en.cppreference.com/w/cpp/language/implicit_conversion#Integral_promotion
+                                // for more about default argument promotion.
+                                let type_name = if let Some(data_type) =
+                                    self.annotation_map.get_type(parameter, self.index)
+                                {
+                                    match &data_type.information {
+                                        DataTypeInformation::Float { .. } => get_bigger_type(
+                                            data_type,
+                                            self.index.get_type_or_panic(LREAL_TYPE),
+                                            self.index,
+                                        )
+                                        .get_name(),
+                                        DataTypeInformation::Integer { .. }
+                                            if !&data_type.information.is_bool() =>
+                                        {
+                                            get_bigger_type(
+                                                data_type,
+                                                self.index.get_type_or_panic(DINT_TYPE),
+                                                self.index,
+                                            )
+                                            .get_name()
+                                        }
+                                        _ => type_name,
+                                    }
+                                } else {
+                                    // might be unreachable? at this point, should always be Some(type)
```

Comment:
these situations should lead us here (we couldnt determine a type when visiting `parameter`):
- we didnt visit `parameter` yet - we visit them at 1299 so this is not the case
- we did visit it, but we could not find a type because the expression  may be bullshit? e.g. (x OR 'hello') or an undefined reference ... (I'm not 100% sure if we would annotate VOID, or annotate nothing)
- I think your strategy is good, lets just keep the original type we guessed so far - probably validation will later cancel this build.

-->  maybe lets debug such a situation with an undefined reference and see whether we hit this line and then adapt the comment accordingly.

Reply:
i have played around with a few different expressions in variadic functions, however I never managed to get into the else-statement. however, I agree that it is probably best to just default to the original type and let the validator handle any unforeseen errors. I will replace the comment here to reflect that